### ISSUE TITLE
Java: `Type.getErasure()` erroneously has `Object` as result on some databases

### ISSUE THREAD

**Issue by Marcono1234** [github/codeql#11264]
### Version
CodeQL CLI 2.11.3

### Description
It looks like for the Java databases of some projects `Type.getErasure()` erroneously has `Object` as result in addition to the expected result. For example this query:
```ql
import java

from ClassOrInterface c
where count(c.getErasure()) > 1
select c, c.getErasure()
```
Has the following results:
|Class|Erasure|
|-|-|
|String|Object|
|String|String|

This is wrong because `Object` is not the erasure type of `String`. Also in general there should always only be 1 erasure type as result for that predicate [^1].

This does not affect all databases and only seems to affect a few specific classes (e.g. `String`), which also differ between databases:

- Class: `java.lang.String`
  Projects:
  - https://github.com/quarkusio/quarkus/commit/967d3497b585af137e69a980c26ba5af30ab84dc
- Classes: `java.lang.String`, `com.android.build.api.variant.impl.DirectoryEntry`, `com.android.manifmerger.ManifestSystemProperty`
  Projects:
  - https://github.com/raphw/byte-buddy/commit/9a818565255791fed4d88c369a4dd3938693ff4c
- Class: `org.apache.logging.log4j.core.async.RingBufferLogEvent`
  Projects:
  - alibaba/arthas
  - alibaba/nacos
  - apache/activemq
  - apache/flink
  - apache/shiro
  - apache/skywalking
  - apache/storm
  - brettwooldridge/HikariCP
  - dropwizard/metrics
  - elastic/logstash
  - google/gson
  - hazelcast/hazelcast
  - hibernate/hibernate-orm
  - micrometer-metrics/micrometer
  - neo4j/neo4j

The following is a reduced version of the `getErasure()` implementation which already allows reproducing this:
```ql
import java

private Type erase(Type t) {
  result = t.(Class).getSourceDeclaration()
  or
  result = erase(t.(BoundedType).getFirstUpperBoundType())
}

from ClassOrInterface c
where count(erase(c)) > 1
select c, erase(c)
```

Interestingly `t.(BoundedType).getFirstUpperBoundType()` seems to have an effect here, even though `ClassOrInterface` and `BoundedType` are incompatible QL classes.

[^1]: There appear to be some other types (wildcards and type variables?) which have more than one result for `getErasure()` as well, but that might be a different issue.

**Comment by aschackmull** [comment]
@smowton This sounds like the same problem that you were investigating a while back with parameterised types having unknown type arguments which were then mistaken for type variables?

**Comment by smowton** [comment]
Certainly possible -- taking a look at this one is on my medium-term backlog

**Comment by smowton** [comment]
This is fixed as of CodeQL 2.12.5. Databases now contain more orderly stub classes when a class is missing from the classpath, which have an entry in the `classes_or_interfaces` table but no other information. 

They may be misclassified as a `Class` rather than an `Interface` (and similarly may not appear as an `Annotation`, `Enum` or other specialised variety of class), since on account of not having the `.class` file to hand we don't know anything about it except for its name. There may also be other unexpected behaviours regarding such stub classes (for example, their not having constructors, appearing as annotations despite not appearing to be annotations, and doubtless others).

However, the particular bug that was causing database inconsistencies and the behaviour Marcono reported above has been fixed. A Multiple-Repository Variant Analysis run over the top 1000 Java repositories shows that there are inconsistencies of the kind reported in this issue except where the database was created with CodeQL <= 2.12.4, so I am closing this as fixed.
